### PR TITLE
Patch SSO Authentication in AD-Forest

### PR DESCRIPTION
**Pull request by chrosey**
Here is our working solution for authenticating an user per single-sign-on in an multidomain-scenario,

only sso is affected, and only if a domain is provided

### CODE REVIEW

File: Classes/Library/Authentication.php [hunk 1]
```diff
@@ -95,6 +96,25 @@ public static function ldapAuthenticate($username, $password = null)
             // Set extension configuration from TYPO3 mode (BE/FE).
             static::initializeConfiguration();
 
+
```
Comment (xperseguers) [inline Classes/Library/Authentication.php]: extra useless empty line

File: Classes/Library/Authentication.php [hunk 2]
```diff
@@ -95,6 +96,25 @@ public static function ldapAuthenticate($username, $password = null)
...
+            if (null != $domain) {
```
Comment (xperseguers) [inline Classes/Library/Authentication.php]: `if (!empty($domain))` as we should handle an empty string as well

File: Classes/Library/Authentication.php [hunk 3]
```diff
@@ -95,6 +96,25 @@ public static function ldapAuthenticate($username, $password = null)
...
+                // Domain is set, so check it
+                if (strpos($domain, '.') > -1 ) {
+                    // Format Domain
+                    $domain = "DC=" . implode(',DC=', explode('.', $domain));
```
Comment (xperseguers) [inline Classes/Library/Authentication.php]: rest of the code is using single quote => `$domain = 'DC=' ...`

File: Classes/Library/Authentication.php [hunk 4]
```diff
@@ -95,6 +96,25 @@ public static function ldapAuthenticate($username, $password = null)
             // Set extension configuration from TYPO3 mode (BE/FE).
             static::initializeConfiguration();
 
+
+            if (null != $domain) {
+                // Domain is set, so check it
+                if (strpos($domain, '.') > -1 ) {
```
Comment (xperseguers) [inline Classes/Library/Authentication.php]:
```
if (strpos($domain, '.') !== false) {
```

File: Classes/Library/Authentication.php [hunk 5]
```diff
@@ -95,6 +96,25 @@ public static function ldapAuthenticate($username, $password = null)
...
+                    // Format Domain
+                    $domain = "DC=" . implode(',DC=', explode('.', $domain));
+                }
+                $domain = strtolower($domain);
+
+                $configDomain = strtolower(static::$config['users']['basedn']);
+                $configDomain = substr($configDomain, strpos($configDomain, 'dc'));
+
+                if ($domain != $configDomain) {
```
Comment (xperseguers) [inline Classes/Library/Authentication.php]: strict comparison `!==`

File: Classes/Library/Authentication.php [hunk 6]
```diff
@@ -95,6 +96,25 @@ public static function ldapAuthenticate($username, $password = null)
...
+                    // Domain does not match, stop validating here
+                    static::getLogger()->notice(sprintf('User Domain "%s" missmatches Configuration Domain "%s"', $domain, $configDomain));
```
Comment (xperseguers) [inline Classes/Library/Authentication.php]: missmatches => mismatches (only one "s")

File: Documentation/Index.rst [hunk 7]
```diff
@@ -57,7 +57,7 @@ LDAP / SSO Authentication
 	**Other version of this document**
 
 	:PDF:
-		https://docs.typo3.org/typo3cms/extensions/ig_ldap_sso_auth/3.3.0/_pdf/
+		https://docs.typo3.org/typo3cms/extensions/ig_ldap_sso_auth/3.3.1/_pdf/
```
Comment (xperseguers) [inline Documentation/Index.rst]: This is not related to the proposed change

Comment (chrosey) [reply Documentation/Index.rst]: Sorry, that one was not intended

**Comment by xperseguers** [comment]
Resolved as 135f1a3f4dbf259c95eae3db5ae182d99c7a5b26
